### ISSUE TITLE
Characterization and propagation of job task type

### ISSUE THREAD

**Issue by vkuznet** [dmwm/WMCore#10604]
**Impact of the new feature**
The monitoring classAds parser needs to extract certain fields from a task name. As such it should follow common rules/convention used in Task Name assignments. If Task Name is constructed without any rules it is not possible to construct generic parser.

The issue is extensively discussed in [CMSMONIT-394](https://its.cern.ch/jira/browse/CMSMONIT-394) and [MM channel](https://mattermost.web.cern.ch/cms-o-and-c/channels/cpu-efficiency).

**Is your feature request related to a problem? Please describe.**
Yes, for monitoring purposes we need specific rules to follow.

**Describe the solution you'd like**
This issue has been divided in categories:
- **Short term**: Improve current logic at the monit level, to account for new string matches. This is already done. See:
https://github.com/dmwm/cms-htcondor-es/pull/206

- **Mid term**: Implement this using the cmsDriver command line arguments
- [x] https://github.com/dmwm/WMCore/issues/11711
- [x] https://github.com/dmwm/WMCore/issues/11712

- **Long term**: If cmssw gives native support for the physics types, use this approach. 
#11713 was created to keep track of this possibility, yet we should focus on the mid term solution, which is in control of WMCore.

**Describe alternatives you've considered**
So far there is a long if/else constructs which decode Task Name in monitoring parser, this code cannot be generalize.

**Additional context**

**Comment by amaltaro** [comment]
@vkuznet both TaskChain and StepChain naming convention depends solely on the requestors (physics group, McM). The only rules we have is that it has to be string of max length 50, AFAIR.

Perhaps it might be better if you can clarify which information is extracted from the task name, such that we can try to provide those as extra job classads, thus not requiring any fragile parsing of fields and strings.

**Comment by vkuznet** [comment]
@amaltaro your question should be addressed to CMS Monitoring group and not to me. @mrceyhun , @leggerf please provide your input over here.

**Comment by leggerf** [comment]
@amaltaro these are the values of task type we are trying to infer with string matching:

"Cleanup","Merge", "LogCollect", "MINIAOD", "DIGIRECO","GENSIM","DIGI","RECO", "GENSIM", "DataProcessing", "Repack", "Express"

You can see the details of the logic directly in the code [here](https://github.com/dmwm/cms-htcondor-es/blob/master/src/htcondor_es/convert_to_json.py#L1067).

It would be great if we can get these values directly as a ClassAd, rather than doing the parsing

@mrceyhun pls add more details if needed

**Comment by amaltaro** [comment]
@leggerf thank you for your input, Federica. Last time this topic came up, I understood that we had to find a robust way to characterize the job task types, instead of relying on parsing the SubTaskName field (which depends on how McM names the tasks).

So, the question I have here is whether the current logic/mapping is good enough and correctly define the task type? If we are not, then we need to start this discussion from the very first source of data (cmsDriver and McM).

**Comment by leggerf** [comment]
it works most of the time, but as you can see in the linked JIRA ticket, 
not always.

I think the logic should be validated/provided by the data producers, 
not the monitoring team. Can you involve them in the discussion and get 
their input?

**Comment by davidlange6** [comment]
It might also be useful to revisit the set of task types. I have two example use cases
a) gen only jobs.
b) step chains naturally do a combination of things, it would be useful to know what each step in the step chain does.

in both cases, it seems that the truth can most reliably come from a dedicated field in the request (that may or may not exist today) rather than parsing some string.

**Comment by davidlange6** [comment]
otherwise, are the input and output data tiers (for each cmsRun step) available to use in some logic?

**Comment by amaltaro** [comment]
> in both cases, it seems that the truth can most reliably come from a dedicated field in the request (that may or may not exist today) rather than parsing some string.

I think we could adapt how TaskName/StepName is used. But creating a new parameter in the request schema is an option as well.

> otherwise, are the input and output data tiers (for each cmsRun step) available to use in some logic?

it should be available. But what we get almost for free is the input "outputModule".

**Comment by davidlange6** [comment]
> On Mar 10, 2023, at 1:46 PM, Alan Malta Rodrigues ***@***.***> wrote:
> 
> 
> in both cases, it seems that the truth can most reliably come from a dedicated field in the request (that may or may not exist today) rather than parsing some string.
> I think we could adapt how TaskName/StepName is used. But creating a new parameter in the request schema is an option as well.
> otherwise, are the input and output data tiers (for each cmsRun step) available to use in some logic?
> it should be available. But what we get almost for free is the input "outputModule".

Definitely you need both input and output to do something more than semi-reliable.

David

**Comment by davidlange6** [comment]
Dunno with out some experimentation/discussion, but something like

Preprocess - ignore dqm outputs unless its the only output

a) find all merge and logcollects

b) find all prompt reco and express 

Make a list of an all possible tiers and combinations of them -- well, find a list of all such combinations used

Make a matrix from these  and lets fill it out - should not be difficult to crowd source that.

**Comment by amaltaro** [comment]
Bugra, Sunil and I had a discussion on it today and it looks like our best alternative to have an automated way to define the physics process type being executed in a job would be by parsing the PSet job configuration - the cmsDriver command line - and extract this information from the `--step XXX` pattern. An example PSet job configuration file is:
https://cmsweb.cern.ch/couchdb/reqmgr_config_cache/b5800d7a387fa74d27c957e2ddbd4853/configFile

A few other remarks based on this cmsDriver step are:
* there might be a chain of steps and some are not really relevant. We would have to define a list of steps that we want to match, and discard anything that does not match that, e.g.: the only relevant step for `--step DIGI,L1,DIGI2RAW,HLT:@relval2021` would be DIGI. Bugra/Sunil would have to provide us with that list though
* for StepChains, we would probably have to have a comma separated string with all the physics process type contained in the job. But to be checked if it would be Okay with the Monitoring team.

The following are still open questions and/or need further investigation:
* are we able to process every PSet configuration in a workflow and collect the relevant information once?
* how do we propagate it all the way to the job? do we persist it in the workflow description?
* how do we define the physics process type if `--step` does not match any of the expected values?
* anything else that I might be missing?

**Comment by khurtado** [comment]
This sounds good! We can use condor chiro to create the classad from the
execution.

If the information in --step does not match any criteria, should we throw a
warning and tag it as "unknown",  should we first try to use the name
matching logic as a fallback before giving up, or should we go with
"unknown" and create an additional classad with the whole --step arguments
for future debugging (or further identification of such args later)?

**Comment by amaltaro** [comment]
I don't yet have all the implementation details, but the way I see it is:
1. something in ReqMgr2 defines the physics type (still to see if the current schema could accommodate it somehow
2. in the agent, JobCreator takes the relevant request parameter and propagate it down to the job dictionary
3. JobSubmitter consumes this information and creates the correct condor classad (likely a new one)
4. if there is no physics type matching any of the expected physics task type, we default it to "unknown"

@khurtado could you please provide a list of the current supported task types? We will need to cross-check that with a list that Bugra/Sunil will provide (to be searched in the PSet configuration).

**Comment by amaltaro** [comment]
Moving this one from Q1 to Q2, as it is still a work in progress item.

**Comment by amaltaro** [comment]
@khurtado Kenyi, we discussed this in one of the weekly meetings, but I see we didn't update this issue. If my memory does not fail me, what we agreed was:
1) to provide a temporary fix to the CMS monitoring script, such that we can deploy it today or in the coming days
2) then in parallel, to keep working on this long term task characterization development.

**Comment by khurtado** [comment]
@amaltaro Item 1 in your comment has now been merged in the following repository:

https://github.com/dmwm/cms-htcondor-es/pull/206

**Comment by khurtado** [comment]
@amaltaro Coming back to this, the supported task types are:
["Production", "Processing", "LogCollect", "Skim", "Merge", "Harvesting", "Cleanup", "Repack", "Express"]
For "Processing", we return "DataProcessing", for "Production", we guess the physics task type:

physics task types: ["GENSIM", "GEN", "DIGI", "RECO", "DIGIRECO", "MINIAOD"]

We would need to find the equivalents for the physics task types in the configFile:
https://cmsweb.cern.ch/couchdb/reqmgr_config_cache/b5800d7a387fa74d27c957e2ddbd4853/configFile

What are the equivalents? 

For stepChains, do we get all the configFile and create a list? E.g.::
https://cmsweb.cern.ch/reqmgr2/fetch?rid=request-pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-04624__v1_T_220629_130256_3425

**Comment by khurtado** [comment]
Related Issues:

cmsDriver feature request: 
https://github.com/cms-sw/cmssw/issues/42450

Help on characterization of job physics task types
https://github.com/cms-sw/cmssw/issues/42587

**Comment by amaltaro** [comment]
@khurtado Kenyi, it looks like we covered everything in this meta-issue and I am closing this out. Please do reopen it if there is anything that is still missing. Thanks!

**Comment by khurtado** [comment]
@amaltaro We still needed a CMS monit ticket in order to use this new feature

I have created such ticket and it can be found below. This should be a trivial change for them
https://github.com/dmwm/cms-htcondor-es/issues/215
